### DISCUSSION TITLE
Unable to pass Dynamic parameter value to a map function in CouchDB

### DISCUSSION THREAD

**Discussion by PreethiVuchuru27916** [apache/couchdb#4134] [General]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description 
I want to pass a dynamic parameter value to the Couchdb map function. But it is failing when additional parameters except doc are passed. Please specify If there is a way to pass a dynamic value to the map function parameter in CouchDB to do some processing inside the function.

[NOTE]: # ( Describe the problem you're encountering. )
[TIP]:  # ( Do NOT give us access or passwords to your actual CouchDB! )

## Steps to Reproduce
[1] Create a design document and Create a view
[2] Add the following code to the map function.

```javascript
function (doc, employees) {//where employees is a array inside a document
  if (doc[employees] && Array.isArray(doc[employees])) {
    result = [];
    doc[employees].forEach(function (employee) {
      result.push(employee.empId);
    });
    emit("empIds", result);
  }
}
```

[NOTE]: # ( Include commands to reproduce, if possible. curl is preferred. )
```
curl "http://admin:adminpw@localhost:5984/ri-prototype/_design/OrganizationInfo/_view/GetEmployees?employees=\"employees"\""
```

## Expected Behaviour
To pass the employees variable to the view as a parameter.
[NOTE]: # ( Tell us what you expected to happen. )
I wanted to have dynamic value to be passed to the map function of a view in CouchDB. 
## Your Environment

[TIP]:  # ( Include as many relevant details about your environment as possible. )
[TIP]:  # ( You can paste the output of curl http://YOUR-COUCHDB:5984/ here. )

* CouchDB version used: CouchDB v. 3.2.2
* Browser name and version: Safari Version 15.3
* Operating system and version: Mac OS 

## Additional Context

[TIP]:  # ( Add any other context about the problem here. )

**Comment by big-r81** [comment]
Hi Preethi,

this will not work! You cannot change the method signature.

Ich vermutet, du möchtest dir alle Angestellten auflisten lassen. Dazu könntest du zum Beispiel in deinen Datensätzen ein Attribute `type` mit dem Wert `employee` einfügen, z.Bsp.:
```
{
  "_id": ".....",
  "firstName": "Preethi",
  "lastName": "Vuchuru",
  "address": "...",
  "type": "employee"
}
```

If you now want to list all employees, you could use this map function (only an example:

```
function(doc) {
    if (doc.type == "employee") {
        emit(doc.LastName, {FirstName: doc.firstName, Address: doc.address});
    }
}
```
This will emit something like this, if you call your view:
```
$db/_design/OrganizationInfo/_view/GetEmployees

{"id":"doc2","key":"Vuchuru","value":{"FirstName":"Preethi","Address":"..."}}
```
This would show you all of your employees order by their last name in ascending order.

You can read more about views in our docs, e.g. https://docs.couchdb.org/en/stable/ddocs/views/index.html

**Reply by PreethiVuchuru27916** [reply]
Thanks Ronny! Is there any way I can pass a dynamic value to be validated against the if condition? 
Is there any notation where I don't have to specify the employee as a hard coded value? For example,
```javascript 
function(doc) {
    if (doc.type == DYNAMIC_VALUE) { // rather than if (doc.type == "employee") {
        emit(doc.LastName, {FirstName: doc.firstName, Address: doc.address});
    }
}
```
So the above function can create a view based on the dynamic value set somewhere. If the dynamic value passed is employee, then it will construct a employee view with key and value pairs. If the dynamic value passed is company, then it will construct a company view with key and value pairs. Can you please confirm if this dynamic nature is possible in a view or if the views needs to be constructed with prior planning and hardcoding in a view is considered normal?

**Comment by big-r81** [comment]
Hi Preethi,

no this is not possible. You can create multiple design documents or you can create multiple views within one design document. So you can construct a view for a company and a view for employees. I don't think it's a good idea to mix those in one ddoc, because they have different data in it.

An other way to do a "dynamic" query is to use the `GET` parameter `startkey`and `endkey` in combination with `include_docs`. You can look at our [docs](https://docs.couchdb.org/en/stable/ddocs/views/collation.html#views-collation) for more infos and examples.

But I'm only guessing at the moment, so can you explain to me what exactly you want to do?

**Comment by PreethiVuchuru27916** [comment]
I have the following document inside couchdb. Now I want to retrieve value for batters.batter.id == 1003 when the type specified by the user is batters. And I want to retrieve value for toppings.id == 5001 when the type specified by the user is topping. Can this be possible without using index in the query? Point to be noted is that : the json structure can be nested as much as possible and still the dynamic query needs to be constructed.
``` json
{
  "items":{
    "item":[
      {
        "id":"0001",
        "batters":{
          "batter":[
            {
              "id":"1003",
              "type":"Blueberry"
            },
            {
              "id":"1004",
              "type":"Devil's Food"
            }
          ]
        },
        "topping":[
          {
            "id":"5001",
            "type":"None"
          },
          {
            "id":"5002",
            "type":"Glazed"
          }
        ]
      }
    ]
  }
}
```

**Comment by big-r81** [comment]
So if you "search" for `batters`, you want to get `Blueberry` as result and if you search for `topping` you want to get `None` as result or do I understand this wrong? Can you give an example `result` or a more detailed example?
